### PR TITLE
Allow async nodes to be canceled by using a context

### PR DESCRIPTION
**Pull request by jbcpollak**
This PR allows async nodes to be canceled and introduces a (messy) unit test to verify this works correctly.

Some other refactoring to support other features and more idiomatic use as well

### CODE REVIEW

File: .devcontainer/devcontainer.json [hunk 1]
```diff
@@ -11,7 +11,9 @@
 				"EditorConfig.EditorConfig"
 			]
 		}
-	}
+	},
+
+	"mounts": ["type=bind,source=/dev/bus/usb,target=/dev/bus/usb"]
```
Comment (jbcpollak) [inline .devcontainer/devcontainer.json:16]:
```suggestion
	}
```

[the remark this answers is not in the document]
Comment (jbcpollak) [reply .devcontainer/devcontainer.json]: oops, wrong project's dev container 🤦

File: common/composite/random_selector.go [hunk 2]
```diff
@@ -22,11 +23,11 @@ type randomSelector[Blackboard any] struct {
 func (s *randomSelector[Blackboard]) Enter(bb Blackboard) {}
 
 // Tick ...
-func (s *randomSelector[Blackboard]) Tick(bb Blackboard, evt core.Event) core.NodeResult {
+func (s *randomSelector[Blackboard]) Tick(bb Blackboard, ctx context.Context, evt core.Event) core.NodeResult {
 	rand.Seed(time.Now().UnixNano())
```
[the remark this answers is not in the document]
Comment (jbcpollak) [reply common/composite/random_selector.go:27]: I saw a warning about this in VSCode, I'll address it at some point, but its existing code and not related to this PR

File: common/decorator/async_delayer.go [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+func (e DelayerFinishedEvent) TargetNodeId() uuid.UUID {
+	return e.targetNodeId
+}
+
+func (d *asyncdelayer[Blackboard]) doDelay(ctx context.Context, enqueue core.EnqueueFn) error {
+	t := time.NewTimer(d.delay)
+	select {
+	case <-ctx.Done():
+		t.Stop()
+		return fmt.Errorf("Interrupted")
```
[the remark this answers is not in the document]
Comment (jbcpollak) [reply common/decorator/async_delayer.go:67]: all makes sense, 👍

File: common/decorator/delayer.go [hunk 4]
```diff
@@ -11,12 +12,12 @@ func Delayer[Blackboard any](params core.Params, child core.Node[Blackboard]) co
 	base := core.NewDecorator("Delayer", params, child)
 	d := &delayer[Blackboard]{Decorator: base}
 
-	ms, err := params.GetInt("ms")
+	delay, err := params.GetInt("delay")
```
[the remark this answers is not in the document]
Comment (jbcpollak) [reply common/decorator/delayer.go:15]: only that this code predates the typeless "Get()" function, I'll update this code so its api matches the async variant as well

File: greenstalk.go [hunk 5]
```diff
@@ -1,23 +1,26 @@
...
 // BehaviorTree ...
 type BehaviorTree[Blackboard any] struct {
+	ctx        context.Context
 	Root       core.Node[Blackboard]
 	Blackboard Blackboard
 	events     chan core.Event
 }
 
 // NewBehaviorTree returns a new BehaviorTree. A data Blackboard
 // to be propagated down the tree each tick is created.
```
Comment (jbcpollak) [inline greenstalk.go:22]:
```suggestion
```

File: greenstalk_test.go [hunk 6]
```diff
@@ -30,60 +31,155 @@ import (
...
 		panic(err)
 	}
 
-	evt := Event{"initial event"}
+	evt := core.DefaultEvent{}
 	go tree.EventLoop(evt)
 
-	time.Sleep(2 * time.Minute)
+	// Wait half the delay and verify no value sent
+	first_halfway, ok := getCount(time.Duration(delay/2) * time.Millisecond)
```
[the remark this answers is not in the document]
Comment (jbcpollak) [reply greenstalk_test.go:150]: yeah, not happy with this, I was hoping the channels would help, but I think yeah a channel that has "definitely failed" indicators is a good call

File: core/update.go [hunk 7]
```diff
@@ -1,15 +1,18 @@
 package core
 
+import "context"
+
 // Update updates a node by calling its Enter method if it is not running,
 // then its Tick method, and finally Leave if it is not still running.
-func Update[Blackboard any](node Node[Blackboard], bb Blackboard, evt Event) NodeResult {
+func Update[Blackboard any](node Node[Blackboard], bb Blackboard, ctx context.Context, evt Event) NodeResult {
```
[the remark this answers is not in the document]
Comment (jbcpollak) [reply core/update.go]: fixed

File: core/node.go [hunk 8]
```diff
@@ -23,20 +41,31 @@ type Node[Blackboard any] interface {
 
 	// Must be implemented by the custom node.
 	Enter(Blackboard)
-	Tick(Blackboard, Event) NodeResult
+	Tick(Blackboard, context.Context, Event) NodeResult
```
[the remark this answers is not in the document]
Comment (jbcpollak) [reply core/node.go]: fixed

File: greenstalk.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (jbcpollak) [reply greenstalk.go]: resolved

File: common/decorator/delayer.go (hunk 4 above)
[the remark this answers is not in the document]
Comment (jbcpollak) [reply common/decorator/delayer.go]: addressed!

**Comment by jbcpollak** [comment]
most comments resolved

File: greenstalk_test.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (fastcat) [reply greenstalk_test.go:137]: You could have the counter node have an `increments chan int` that it (tries to) write to when it fires and this then can select across that and a timeout channel

**Review by nivanov-ati — APPROVED** [review]

**Pull request merged**